Extending DPL input handling and message caching related to [O2 1325]

matthiasrichter opened the pull request:
This is a prototype which also will allow with a few more changes to activate support for an updated O2 message model where split payloads can be transferred in a sequence with one preceding data header.

Has to be finally activated in `DataProcessingDevice` together with a workflow test.

For using this in the workflows together with Data Distribution, also the raw proxy needs to be extended, and for that a better encapsulation of the message model is in development based on the outcome of the prototype test.

matthiasrichter committed:
DPL: Implementing support for multiple payloads per header

This is supporting an upcoming extension of the O2 data model to allow
not only header-payload pairs but also sequences of payloads preceded
by a single header message. Such a sequence can be indicated by setting
splitPayloadIndex == splitpayloadParts with splitpayloadParts > 1.
A change in the version of DataHeader will be discussed, although there
is no change in the members of the header.

There is no change on the InputRecord API, only the payload size of the
current payload is provided in the DataRef object. It needs to be discussed,
which payload size is going to be provided in the data header in case of
a sequence with multiple payloads.

Changes alongside (most of them meanwhile moved to separate commits):
- make MessageSet the owner of FairMQMessage objects in a linear vector,
  enhances performance also for split payload sequences in header-paiload
  format
- generalizing DataRelayer to relay array of messages
- storing the payload message size in the DataRef

Extending and cleaning up DataRelayer test and benchmark
- allocating header stack through transport allocator and avoiding copy
- adding test for split parts in the header-payload scheme
- adding test for split parts in payloads sequence scheme

During the development a performance regression has been observed when
using o2::pmr::getMessage, needs to be investigated.

ktf commented on Framework/Core/include/Framework/PartRef.h:
```diff
@@ -21,9 +21,66 @@ namespace framework
+  }
+
+ private:
+  std::vector<std::unique_ptr<FairMQMessage>> mMessages;
```
I am a bit worried we will start getting vector allocations all over the place even for the (common?) case of a single {header,payload} pair. Is there any way we can avoid that?

matthiasrichter replied:
sharing your concern, we could have the vector in addition to the two standard members and handle it in the access functions. Assuming that an empty vector has less impact in terms of allocation. To have some numbers I will do a benchmark.

ktf replied:
Yes, sounds like a good idea, waiting for the benchmark.

matthiasrichter replied:
This has been fixed, PartRef is deprecated now, MessageSet is using a linear storage for all messages belonging to one cell in the DataRelayer cache

matthiasrichter commented:
I have changed `MessageSet` to keep one vector of messages and an index to access the different parts. `PartRef` becomes almost obsolete, there is one dependencies for the messages created by DPL, like e.g. timers. This makes the `DataRelayer` much faster for the split parts payloads.

from `o2-bench-framework-benchmark-DataRelayer`
```
Run on (8 X 3400 MHz CPU s)
2021-10-07 13:56:35
---------------------------------------------------------------
Benchmark                        Time           CPU Iterations
---------------------------------------------------------------
BM_RelayMessageCreation        192 ns        192 ns    3641912
BM_RelaySingleSlot             926 ns        926 ns     727922
BM_RelayMultipleSlots          996 ns        996 ns     692755
BM_RelayMultipleRoutes        1967 ns       1967 ns     354810
BM_RelaySplitParts            1533 ns       1527 ns     489359
```

EDIT Nov 22: adding the benchmark result for the final version
```
Run on (8 X 3400 MHz CPU s)
2021-11-22 22:33:46
---------------------------------------------------------------
Benchmark                        Time           CPU Iterations
---------------------------------------------------------------
BM_RelayMessageCreation        197 ns        197 ns    3619241
BM_RelaySingleSlot             970 ns        970 ns     722423
BM_RelayMultipleSlots         1000 ns       1000 ns     696814
BM_RelayMultipleRoutes        1999 ns       1999 ns     349712
BM_RelaySplitParts            1725 ns       1715 ns     453292
```

Compared to the original implementation
```
Run on (8 X 3400 MHz CPU s)
2021-10-07 13:58:45
---------------------------------------------------------------
Benchmark                        Time           CPU Iterations
---------------------------------------------------------------
BM_RelayMessageCreation        192 ns        192 ns    3557891
BM_RelaySingleSlot             915 ns        915 ns     747876
BM_RelayMultipleSlots          974 ns        974 ns     697222
BM_RelayMultipleRoutes        1851 ns       1851 ns     379642
BM_RelaySplitParts            4558 ns       4160 ns     188960
```
The first draft of this PR with the message vector inside `PartRef` was even slower (expectedly).

Also the `DataProcessingDevice` is in principle ready to handle the new data format, but I need to finish off the unit test.

matthiasrichter commented:
@ktf I need to find out whats wrong with the unit test in the macos build, lets see if this is a problem also on the other CIs, then need to find out why its working on my Ubuntu installation. But in general the code in `DataProcessingDevice` and `DataRelayer` is ready.

matthiasrichter committed:
Appearently, DPL callback adapters do noy yet support const parameters

Using e.g. RawDeviceService const& in the init callback leads to an exception which
is difficult to understand and it comes from the adaptStateful converters
(_ZNK2o29framework15ServiceRegistry3getEjmNS0_11ServiceKindEPKc+0x2b9)[0x56376fa3f8b1]

alibuild commented:
Error while checking build/O2/o2-dataflow-cs8 for e284c1bcdfde4f60b9bfdb8a1e7d0866ada88363 at 2021-11-03 09:43:
```
sw/BUILD/O2-latest/log
      Start  98: test_Framework_test_SimpleStatefulProcessing01
84/98 Test  #98: test_Framework_test_SimpleStatefulProcessing01 ........   Passed    0.59 sec
      Start  99: test_Framework_test_SimpleStringProcessing
85/98 Test  #99: test_Framework_test_SimpleStringProcessing ............   Passed    0.56 sec
      Start 100: test_Framework_test_SimpleTimer
86/98 Test #100: test_Framework_test_SimpleTimer .......................   Passed    1.53 sec
      Start 101: test_Framework_test_SimpleWildcard
87/98 Test #101: test_Framework_test_SimpleWildcard ....................   Passed    1.67 sec
      Start 102: test_Framework_test_SimpleWildcard02
88/98 Test #102: test_Framework_test_SimpleWildcard02 ..................   Passed    1.74 sec
      Start 103: test_Framework_test_SingleDataSource
89/98 Test #103: test_Framework_test_SingleDataSource ..................   Passed    1.53 sec
      Start 104: test_Framework_test_Task
90/98 Test #104: test_Framework_test_Task ..............................   Passed    0.73 sec
      Start 105: test_Framework_test_ExternalFairMQDeviceWorkflow
91/98 Test #105: test_Framework_test_ExternalFairMQDeviceWorkflow ......   Passed    2.40 sec
      Start 106: test_Framework_test_VariablePayloadSequenceWorkflow
92/98 Test #106: test_Framework_test_VariablePayloadSequenceWorkflow ...   Passed    1.19 sec
      Start 107: test_Framework_test_ProcessorOptions
93/98 Test #107: test_Framework_test_ProcessorOptions ..................   Passed    0.70 sec
      Start 116: test_Framework_benchmark_RawParser
94/98 Test #116: test_Framework_benchmark_RawParser ....................   Passed   12.68 sec
      Start 117: test_Framework_benchmark_DPLRawPageSequencer
95/98 Test #117: test_Framework_benchmark_DPLRawPageSequencer ..........   Passed    7.33 sec
      Start 118: test_Framework_test_CustomGUIGL
96/98 Test #118: test_Framework_test_CustomGUIGL .......................***Not Run (Disabled)   0.00 sec
      Start 119: test_Framework_test_CustomGUISokol
97/98 Test #119: test_Framework_test_CustomGUISokol ....................***Not Run (Disabled)   0.00 sec
      Start 120: test_Framework_test_SimpleTracksED
98/98 Test #120: test_Framework_test_SimpleTracksED ....................***Not Run (Disabled)   0.00 sec

100% tests passed, 0 tests failed out of 93

Label Time Summary:
benchmark    =  20.01 sec*proc (2 tests)
dplutils     =  20.01 sec*proc (2 tests)
framework    =  55.46 sec*proc (89 tests)
workflow     =  38.57 sec*proc (26 tests)

Total Test time (real) =  75.71 sec

The following tests did not run:
	 85 - test_Framework_test_BoostSerializedProcessing (Disabled)
	 88 - test_Framework_test_DanglingInputs (Disabled)
	118 - test_Framework_test_CustomGUIGL (Disabled)
	119 - test_Framework_test_CustomGUISokol (Disabled)
	120 - test_Framework_test_SimpleTracksED (Disabled)
+ set +x
+ [[ ! -n '' ]]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/7110/e284c1bcdfde4f60b9bfdb8a1e7d0866ada88363/build_O2_o2-dataflow-cs8/pretty.html).

alibuild commented:
Error while checking build/O2/o2-cs8 for e284c1bcdfde4f60b9bfdb8a1e7d0866ada88363 at 2021-11-03 04:50:
```
sw/BUILD/O2-latest/log
      Start  98: test_Framework_test_SimpleStatefulProcessing01
84/98 Test  #98: test_Framework_test_SimpleStatefulProcessing01 ........   Passed    0.59 sec
      Start  99: test_Framework_test_SimpleStringProcessing
85/98 Test  #99: test_Framework_test_SimpleStringProcessing ............   Passed    0.69 sec
      Start 100: test_Framework_test_SimpleTimer
86/98 Test #100: test_Framework_test_SimpleTimer .......................   Passed    1.59 sec
      Start 101: test_Framework_test_SimpleWildcard
87/98 Test #101: test_Framework_test_SimpleWildcard ....................   Passed    2.40 sec
      Start 102: test_Framework_test_SimpleWildcard02
88/98 Test #102: test_Framework_test_SimpleWildcard02 ..................   Passed    2.00 sec
      Start 103: test_Framework_test_SingleDataSource
89/98 Test #103: test_Framework_test_SingleDataSource ..................   Passed    1.62 sec
      Start 104: test_Framework_test_Task
90/98 Test #104: test_Framework_test_Task ..............................   Passed    0.77 sec
      Start 105: test_Framework_test_ExternalFairMQDeviceWorkflow
91/98 Test #105: test_Framework_test_ExternalFairMQDeviceWorkflow ......   Passed    2.20 sec
      Start 106: test_Framework_test_VariablePayloadSequenceWorkflow
92/98 Test #106: test_Framework_test_VariablePayloadSequenceWorkflow ...   Passed    1.19 sec
      Start 107: test_Framework_test_ProcessorOptions
93/98 Test #107: test_Framework_test_ProcessorOptions ..................   Passed    0.60 sec
      Start 116: test_Framework_benchmark_RawParser
94/98 Test #116: test_Framework_benchmark_RawParser ....................   Passed   12.67 sec
      Start 117: test_Framework_benchmark_DPLRawPageSequencer
95/98 Test #117: test_Framework_benchmark_DPLRawPageSequencer ..........   Passed    6.95 sec
      Start 118: test_Framework_test_CustomGUIGL
96/98 Test #118: test_Framework_test_CustomGUIGL .......................***Not Run (Disabled)   0.00 sec
      Start 119: test_Framework_test_CustomGUISokol
97/98 Test #119: test_Framework_test_CustomGUISokol ....................***Not Run (Disabled)   0.00 sec
      Start 120: test_Framework_test_SimpleTracksED
98/98 Test #120: test_Framework_test_SimpleTracksED ....................***Not Run (Disabled)   0.00 sec

100% tests passed, 0 tests failed out of 93

Label Time Summary:
benchmark    =  19.61 sec*proc (2 tests)
dplutils     =  19.61 sec*proc (2 tests)
framework    =  67.41 sec*proc (89 tests)
workflow     =  41.99 sec*proc (26 tests)

Total Test time (real) =  87.49 sec

The following tests did not run:
	 85 - test_Framework_test_BoostSerializedProcessing (Disabled)
	 88 - test_Framework_test_DanglingInputs (Disabled)
	118 - test_Framework_test_CustomGUIGL (Disabled)
	119 - test_Framework_test_CustomGUISokol (Disabled)
	120 - test_Framework_test_SimpleTracksED (Disabled)
+ set +x
+ [[ ! -n '' ]]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/7110/e284c1bcdfde4f60b9bfdb8a1e7d0866ada88363/build_O2_o2-cs8/pretty.html).

matthiasrichter commented:
To isolate a problem with a unit test in QualityControl, most of the changes have been committed and merged in separate PRs: #7376, #7493, #7530, #7533, #7558

ktf commented on Framework/Core/src/DataProcessingDevice.cxx:
```diff
@@ -1155,26 +1206,62 @@ bool DataProcessingDevice::tryDispatchComputation(DataProcessorContext& context,
+  // object. The part index within the input needs to be translated into a 2 dim
+  // index to access the payload message i from PartRef object j and we generate
+  // a consecutive map of these tuples
+  std::vector<std::vector<std::pair<size_t, size_t>>> partRefMaps;
```
Can we use a proper struct, rather than a `std::pair<size_t, size_t>`?

matthiasrichter replied:
Fixed

ktf commented:
The error is:

```
/sw/slc7_x86-64/FairRoot/v18.4.2-46/lib
Info in <TUnixSystem::ACLiC>: creating shared library /sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros//sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/TPCFastTransformation/macro/fastTransformQA_C.so
/sw/slc7_x86-64/GCC-Toolchain/v10.2.0-alice2-4/bin/../lib/gcc/x86_64-unknown-linux-gnu/10.2.0/../../../../x86_64-unknown-linux-gnu/bin/ld: cannot open output file /sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros//sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/TPCFastTransformation/macro/fastTransformQA_C.so: No such file or directory
collect2: error: ld returned 1 exit status
Error in <ACLiC>: Executing 'cd "/sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros" ; c++ -fPIC -c -O3 -std=c++17 -Wno-implicit-fallthrough -Wno-noexcept-type -pipe -W -Woverloaded-virtual -fsigned-char -pthread  -I$ROOTSYS/include -I"/sw/slc7_x86-64/pythia/v8304-21/include" -I"/sw/slc7_x86-64/ms_gsl/3.1.0-10/include" -I"/sw/slc7_x86-64/Vc/1.4.1-74/include" -I"/sw/slc7_x86-64/FairRoot/v18.4.2-46/include" -I"/sw/slc7_x86-64/FairMQ/v1.4.43-3/include/fairmq" -I"/sw/slc7_x86-64/asio/v1.19.1-4/include" -I"/sw/slc7_x86-64/FairLogger/v1.10.2-1/include" -I"/sw/slc7_x86-64/fmt/8.0.1-1/include" -I"/sw/slc7_x86-64/GEANT3/v3-7-77/include/TGeant3" -I"/sw/slc7_x86-64/GEANT4_VMC/v5-3-83/include/g4root" -I"/sw/slc7_x86-64/GEANT4_VMC/v5-3-83/include/geant4vmc" -I"/sw/slc7_x86-64/GEANT4_VMC/v5-3-83/include/mtroot" -I"/sw/slc7_x86-64/GEANT4/v10.7.2-7/include" -I"/sw/slc7_x86-64/GEANT4/v10.7.2-7/include/Geant4" -I"/sw/slc7_x86-64/boost/v1.75.0-43/include" -I"/sw/slc7_x86-64/JAliEn-ROOT/0.6.6-84/include" -I"/sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/stage/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/TPCFastTransformation" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/TPCFastTransformation/devtools" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/Common" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Framework/Logger/include" -I"/sw/slc7_x86-64/fmt/8.0.1-1/include" -I"/sw/slc7_x86-64/FairLogger/v1.10.2-1/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/Utils" -I"/sw/slc7_x86-64/ROOT/v6-24-06-37/include" -I"/sw/slc7_x86-64/Vc/1.4.1-74/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/Detectors/TPC/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/simulation/include" -I"/sw/slc7_x86-64/ms_gsl/3.1.0-10/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/Detectors/Common/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/MathUtils/include" -I"/sw/slc7_x86-64/FairRoot/v18.4.2-46/include" -I"/sw/slc7_x86-64/arrow/v5.0.0-alice3-7/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/Constants/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/Headers/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/MemoryResources/include" -I"/sw/slc7_x86-64/FairMQ/v1.4.43-3/include/fairmq" -I"/sw/slc7_x86-64/FairMQ/v1.4.43-3/include" -I"/sw/slc7_x86-64/boost/v1.75.0-43/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Utilities/rANS/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/Utils/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/common/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/Base/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/Reconstruction/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/Field/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Framework/Core/include" -I"/sw/slc7_x86-64/Common-O2/v1.6.0-28/include" -I"/sw/slc7_x86-64/Configuration/v2.6.3-5/include" -I"/sw/slc7_x86-64/libInfoLogger/v2.4.1-1/include" -I"/sw/slc7_x86-64/Monitoring/v3.9.0-1/include" -I"/sw/slc7_x86-64/curl/7.70.0-38/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/CCDB/include" -I"/sw/slc7_x86-64/libjalienO2/0.1.3-11/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Framework/Foundation/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Framework/Foundation/3rdparty/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Utilities/PCG/include" -I"/sw/slc7_x86-64/RapidJSON/v1.1.0-alice2-19/include" -I"/sw/slc7_x86-64/libuv/v1.40.0-14/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/DataFormats/Parameters/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/Types/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Common/SimConfig/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Algorithm/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/TPC/simulation/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/TPC/base/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/Raw/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/TPC/spacecharge/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/Detectors/TPC/reconstruction/include" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/GPUTracking/Definitions" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/GPUTracking/DataTypes" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/GPUTracking/Interface" -I"/sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/GPUTracking" -I"/sw/slc7_x86-64/ROOT/v6-24-06-37/etc/" -I"/sw/slc7_x86-64/ROOT/v6-24-06-37/etc//cling" -I"/sw/slc7_x86-64/ROOT/v6-24-06-37/include/"   -D__ACLIC__ "/sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros/fastTransformQA_C_ACLiC_dict.cxx" ; c++ -O3 "/sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros/fastTransformQA_C_ACLiC_dict.o" -shared   "/lib64/libdl.so" "/lib64/libc.so" "/sw/slc7_x86-64/GCC-Toolchain/v10.2.0-alice2-4/lib64/libgcc_s.so" "/lib64/libm.so" "/sw/slc7_x86-64/GCC-Toolchain/v10.2.0-alice2-4/lib64/libstdc++.so" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libRint.so.6.24" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libCore.so.6.24" "/lib64/libpthread.so.0" "/sw/slc7_x86-64/AliEn-Runtime/v2-19-le-105/lib/libz.so.1" "/lib64/libnss_files.so.2" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libRIO.so" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libThread.so.6.24" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libtbb.so.2" "/lib64/librt.so.1" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libCling.so" "/lib64/libtinfo.so.5" "/usr/lib64/ld-2.17.so" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libMathCore.so.6.24.06" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libImt.so.6.24" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libMultiProc.so.6.24" "/sw/slc7_x86-64/ROOT/v6-24-06-37/lib/libNet.so.6.24" "/sw/slc7_x86-64/AliEn-Runtime/v2-19-le-105/lib/libssl.so.1.0.0" "/sw/slc7_x86-64/AliEn-Runtime/v2-19-le-105/lib/libcrypto.so.1.0.0" -o "/sw/BUILD/7b17cc866b7721c89d903ad45c0e2c405507adeb/O2/compiled_macros//sw/SOURCES/O2/7110-slc7_x86-64/0/GPU/TPCFastTransformation/macro/fastTransformQA_C.so"' failed!

Exit code=1
```

could it be we have too many open files? It looks weird. Anyways, it's most likely unrelated, so merging this.

ktf commented:
Regarding the issue with const / non const, the registry is separate between the two, so that we can avoid that people ask for const services as non-const, however we could indeed extend things so that the opposite is possible.
